src/qt/walletcontroller.cpp

```diff
@@ -140,23 +144,151 @@ void WalletController::removeAndDeleteWallet(WalletModel* wallet_model)
     delete wallet_model;
 }
 
+WalletControllerActivity::WalletControllerActivity(WalletController* wallet_controller, QWidget* parent_widget)
+    : QObject(wallet_controller)
+    , m_wallet_controller(wallet_controller)
+    , m_parent_widget(parent_widget)
+{
+}
+
+WalletControllerActivity::~WalletControllerActivity()
+{
+    delete m_progress_dialog;
+}
+
+void WalletControllerActivity::showProgressDialog(const QString& label_text)
+{
+    m_progress_dialog = new QProgressDialog(m_parent_widget);
+
+    m_progress_dialog->setLabelText(label_text);
+    m_progress_dialog->setRange(0, 0);
+    m_progress_dialog->setCancelButton(nullptr);
+    m_progress_dialog->setWindowModality(Qt::ApplicationModal);
+    GUIUtil::PolishProgressDialog(m_progress_dialog);
+
+    connect(m_progress_dialog, &QObject::destroyed, [this] {
+        m_progress_dialog = nullptr;
+    });
+}
+
+CreateWalletActivity::CreateWalletActivity(WalletController* wallet_controller, QWidget* parent_widget)
+    : WalletControllerActivity(wallet_controller, parent_widget)
+{
+    m_passphrase.reserve(MAX_PASSPHRASE_SIZE);
+}
+
+CreateWalletActivity::~CreateWalletActivity()
+{
+    delete m_create_wallet_dialog;
+    delete m_passphrase_dialog;
+}
 
-OpenWalletActivity::OpenWalletActivity(WalletController* wallet_controller, const std::string& name)
-    : m_wallet_controller(wallet_controller)
-    , m_name(name)
-{}
+void CreateWalletActivity::askPasshprase()
+{
+    m_passphrase_dialog = new AskPassphraseDialog(AskPassphraseDialog::Encrypt, m_parent_widget, &m_passphrase);
+    m_passphrase_dialog->show();
 
-void OpenWalletActivity::open()
+    connect(m_passphrase_dialog, &QObject::destroyed, [this] {
+        m_passphrase_dialog = nullptr;
+    });
+    connect(m_passphrase_dialog, &QDialog::accepted, [this] {
+        createWallet();
+    });
+    connect(m_passphrase_dialog, &QDialog::rejected, [this] {
+        Q_EMIT finished();
+    });
+}
+
+void CreateWalletActivity::createWallet()
 {
-    std::string error, warning;
-    std::unique_ptr<interfaces::Wallet> wallet = m_wallet_controller->m_node.loadWallet(m_name, error, warning);
-    if (!warning.empty()) {
-        Q_EMIT message(QMessageBox::Warning, QString::fromStdString(warning));
+    showProgressDialog(tr("Creating Wallet <b>%1</b>...").arg(m_create_wallet_dialog->walletName().toHtmlEscaped()));
+
+    std::string name = m_create_wallet_dialog->walletName().toStdString();
+    uint64_t flags = 0;
+    if (m_create_wallet_dialog->disablePrivateKeys()) {
+        flags |= WALLET_FLAG_DISABLE_PRIVATE_KEYS;
     }
-    if (wallet) {
-        Q_EMIT opened(m_wallet_controller->getOrCreateWallet(std::move(wallet)));
-    } else {
-        Q_EMIT message(QMessageBox::Critical, QString::fromStdString(error));
+    if (m_create_wallet_dialog->blank()) {
+        flags |= WALLET_FLAG_BLANK_WALLET;
     }
+
+    QTimer::singleShot(500, worker(), [this, name, flags] {
```

Comment:
Nit: would be nice to hoist the 4 occurrences of this magic value (500) to a helpfully named constant.